app/scripts/lib/rpc-method-middleware/handlers/ethereum-chain-utils.js

```diff
@@ -163,30 +194,92 @@ export async function switchChain(
     setActiveNetwork,
     endApprovalFlow,
     getCaveat,
-    requestPermittedChainsPermission,
-    grantPermittedChainsPermissionIncremental,
+    requestPermissionApprovalForOrigin,
```

Comment:
Previously the `requestPermittedChainsPermission` hook was responsible for both 1) getting user consent, and 2) granting permissions. They were co-located in a handler in `metamask-controller.js`. So we _kinda_ had consent and authorization happening in the same place (though this was degraded when the `grantPermittedChainsPermissionIncremental` hook was created and added here).

It seems that now `switchChain` is responsible for _both_ requesting permission and granting that permission, rather than the hooks .We're now relying on more parts of the codebase to handle permissions appropriately. This makes it more likely we'll introduce a bug where a permission is granted when it shouldn't be, and it makes it more difficult to audit what happened if we are trying to investigate a report along those lines.

In terms of limiting the scope of code that can grant permissions, this seems like a step in hte wrong direction. Though admittedly we weren't in a great place before either, as `metamask-controller.js` is a monster, and this hooks system is itself kinda complicated.

What are your thoughts on creating "request and grant permission" hooks, so that the sensitive code that is responsible for safeguarding permissions is more centralized?

Reply:
oof... that also begs the question. should `requestPermissionApprovalForOrigin` be brought into MMC as well and also contain the

```
    let newCaveatValue = {
      requiredScopes: {},
      optionalScopes: {},
      isMultichainOrigin: false,
    };

    if (isSnapId(origin)) {
      newCaveatValue.optionalScopes = {
        'wallet:eip155': {
          accounts: [],
        },
      };
    } else {
      newCaveatValue = setPermittedEthChainIds(
        newCaveatValue,
        legacyApproval.approvedChainIds,
      );
    }

    newCaveatValue = setEthAccounts(
      newCaveatValue,
      legacyApproval.approvedAccounts,
    );
```
CAIP-25 permission forming logic too

Reply:
https://github.com/MetaMask/metamask-extension/pull/29213